tests/backends/oracle/test_utils.py

```diff
@@ -0,0 +1,25 @@
+import unittest
+
+from django.db import connection, models
+from django.test import SimpleTestCase
+
+
+@unittest.skipUnless(connection.vendor == 'oracle', 'Oracle tests')
+class InsertReturnParameterTest(SimpleTestCase):
+
+    def test_auto_field(self):
+        from django.db.backends.oracle.utils import InsertReturnParameter
+
+        for field_name, db_type in InsertReturnParameter.types.items():
+            field_class = getattr(models, field_name)
+            with self.subTest(field_name=field_name, db_type=db_type):
+                class TestModel(models.Model):
+                    custom_pk = field_class(primary_key=True)
+
+                    class Meta:
+                        app_label = 'oracle_%s_test' % field_name
+
+                param = InsertReturnParameter(TestModel._meta.pk)
+                fn = param.get_field_name(param.field)
+                self.assertEquals(fn, field_name)
+                self.assertEquals(param.types.get(field_name), db_type)
```

Comment:
I think we should add real-life tests, e.g. `DateTimeField` filled by database trigger
```diff
diff --git a/tests/backends/models.py b/tests/backends/models.py
index 1fa8d44e63..a2c8616cc6 100644
--- a/tests/backends/models.py
+++ b/tests/backends/models.py
@@ -5,6 +5,10 @@ from django.contrib.contenttypes.models import ContentType
 from django.db import models
 
 
+class NonIntegerAutoField(models.Model):
+    creation_datetime = models.DateTimeField(primary_key=True)
+
+
 class Square(models.Model):
     root = models.IntegerField()
     square = models.PositiveIntegerField()
diff --git a/tests/backends/oracle/tests.py b/tests/backends/oracle/tests.py
index a0d49854d9..e25f88cf54 100644
--- a/tests/backends/oracle/tests.py
+++ b/tests/backends/oracle/tests.py
@@ -1,3 +1,4 @@
+import datetime
 import unittest
 
 from django.db import connection
@@ -5,7 +6,7 @@ from django.db.models.fields import BooleanField, NullBooleanField
 from django.db.utils import DatabaseError
 from django.test import TransactionTestCase
 
-from ..models import Square
+from ..models import NonIntegerAutoField, Square
 
 
 @unittest.skipUnless(connection.vendor == 'oracle', 'Oracle tests')
@@ -95,3 +96,23 @@ class TransactionalTests(TransactionTestCase):
             self.assertIn('ORA-01017', context.exception.args[0].message)
         finally:
             connection.settings_dict['PASSWORD'] = old_password
+
+    def test_non_integer_auto_field(self):
+        with connection.cursor() as cursor:
+            # Create trigger that fill non-integer auto field.
+            cursor.execute("""
+                CREATE OR REPLACE TRIGGER "TRG_FILL_CREATION_DATETIME"
+                BEFORE INSERT ON "BACKENDS_NONINTEGERAUTOFIELD"
+                FOR EACH ROW
+                BEGIN
+                    :NEW.CREATION_DATETIME := SYSTIMESTAMP;
+                END;
+            """)
+        try:
+            NonIntegerAutoField._meta.auto_field = NonIntegerAutoField.creation_datetime
+            obj = NonIntegerAutoField.objects.create()
+            self.assertIsNotNone(obj.creation_datetime)
+            self.assertIsInstance(obj.creation_datetime, datetime.datetime)
+        finally:
+            with connection.cursor() as cursor:
+                cursor.execute('DROP TRIGGER "TRG_FILL_CREATION_DATETIME"')
```
It's a bit hacky but it shows a use case.